### DISCUSSION TITLE
Previews from command line no longer work

### DISCUSSION THREAD

**Discussion by labo-China** [Textualize/textual#2945] [Q&A] [answered]
Please give a brief but clear explanation of the issue. If you can, include a complete working example that demonstrates the bug. **Check it can run without modifications.**

When launching previews from command line (e.g. calling `textual easing`) , it ends up raising exceptions as following:
> Traceback (most recent call last):
  File "/home/labo/prog/NCSbox/venv/bin/textual", line 8, in <module>
    sys.exit(run())
             ^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/labo/prog/NCSbox/venv/lib/python3.11/site-packages/textual_dev/cli.py", line 109, in easing
    from textual.cli.previews import easing
ModuleNotFoundError: No module named 'textual.cli'

From exception prompts we can know that this is due to misplacement of files in refactoring dev tools, which means `textual-dev/previews` should be placed in `textual/cli/previews`. After moving files mannually, everything works fine.

I don`t know whether the preview feature should be preserved in textual or move into textual-dev. I would like to create a PR to fix it if I know where it should be.

**Comment by labo-China** [comment]
By the way, the `BORDER` attribute in constants.py is also missing when running `textual borders` or `textual easing`

**Comment by TomJGooding** [comment]
What version of `textual` and `textual-dev` are you running? [It looks like this was already  moved a few weeks ago.](https://github.com/Textualize/textual-dev/commits/main)

**Comment by davep** [comment]
As @TomJGooding points out above, it would be helpful if you could say what versions you're using, and also how you have your Python and Textual development environments set up. On the surface it would seem that you have an environment that has some mixed up versions.

Here's me setting up a brand new virtual environment with `textual` and `textual-dev` and running the samples just fine:

https://github.com/Textualize/textual/assets/28237/3c5873f7-8f01-4671-8f59-a11db7301a3c

Could you perhaps explain how you go about doing this?

**Comment by labo-China** [comment, marked as answer]
Oops, it was my mistake. I used an mirror site for PIP to download and upgrade packages for faster speed. It turns out that the mirror site was kind of out-dated and gave me `textual 0.29.0` and `textual-dev 0.0.2`. After I unset the mirror site and did a upgrade, the out-dated `textual-dev` package is up-to-date. Sorry for inconvenience and bothering (T_T).
